### ISSUE TITLE
Cannot access docker services after adjusting MTU on private interface 

### ISSUE THREAD

**Issue by cookandy** [docker/docker#26473]
Hi,

I am unable to access services in docker containers after setting the MTU for my private network to `1420`.  More info below...

Some background:

- 2 machines - each running `Ubuntu 16` with `Docker 12`
- Each machine has 2 interfaces - `eth0` (public) and `eth1` (private)
- IPSec (transport mode) is enabled on `eth1`

Because of IPSec overhead, if I run a Tomcat container on Machine A, the only way I can get to it from Machine B is to set the MTU of `eth1` to `1420` on machine B.  **Note, I don't have to adjust any MTU values on the Tomcat host (Machine A) in order for this to work - only Machine B.**  

The problem arises when I try to adjust the MTU value on Machine A.  Because Machine A also needs to access Tomcat services on Machine B, I must adjust the MTU to `1420` just as I did on Machine B.  However, as soon as I do this, I can no longer access the Tomcat container on Machine A anymore (from Machine B).

This is easily reproducible from a default install (with IPSec already established):

1. Leave default MTU of `1500` on Machine A - start Tomcat
2. Leave default MTU of `1500` on Machine B, try to access Tomcat on Machine A using curl - it fails
3. Adjust the MTU of `eth1` to `1420` on Machine B:  `ifconfig eth1 mtu 1420`
4. Try curl again from Machine B - **it works!**
5.  Adjust MTU of `eth1` to `1420` on Machine A:  `ifconfig eth1 mtu 1420`
6. Try curl again from Machine B - **it fails**
7. Try to set back MTU to `1500` on Machine A:  `ifconfig eth1 mtu 1500`
8. Try curl again - **it fails**
9.  Restart docker daemon on Machine A:  `systemctl restart docker`
10. Try curl again - it fails
11. Bring down and up `eth1` on Machine A:  `ifdown eth1 && ifup eth1`
12. Try curl again - **it succeeds** (however MTU is back to `1500`)

How can I set the MTU to `1420` on `eth1` on **BOTH** Machine A and Machine B and still get to the docker services from each?  I have tried to adjust the `docker0` MTU to `1420` on both hosts, but it doesn't seem to make a difference.  

**Output of `docker version`:**

```
Docker version 1.12.1, build 23cf638
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 11
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge host null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-36-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 488.5 MiB
Name: andy-test-01
ID: Z3E6:FAAV:3SGL:O6MA:BYNB:23FJ:QBQX:QSHZ:NRFM:3TPD:CZG4:U6GD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
DigitalOcean

**Steps to reproduce the issue:**
1. Leave default MTU of `1500` on Machine A - start Tomcat
2. Leave default MTU of `1500` on Machine B, try to access Tomcat on Machine A using curl - it fails
3. Adjust the MTU of `eth1` to `1420` on Machine B:  `ifconfig eth1 mtu 1420`
4. Try curl again from Machine B - **it works!**
5.  Adjust MTU of `eth1` to `1420` on Machine A:  `ifconfig eth1 mtu 1420`
6. Try curl again from Machine B - **it fails**
7. Try to set back MTU to `1500` on Machine A:  `ifconfig eth1 mtu 1500`
8. Try curl again - **it fails**
9.  Restart docker daemon on Machine A:  `systemctl restart docker`
10. Try curl again - it fails
11. Bring down and up `eth1` on Machine A:  `ifdown eth1 && ifup eth1`
12. Try curl again - **it succeeds** (however MTU is back to `1500`)

**Describe the results you received:**
Cannot access Tomcat container after changing MTU to `1420` on host running Tomcat container. 

**Describe the results you expected:**
Should be able to access all Docker containers with `eth1` set to `1420` on both hosts.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by aboch** [comment]
@cookandy Does it work if you do not use docker containers for your tomcat app ?
IOW, if you start the tomcat servers on host A and B (not in a container) does it work or do you see the same problem ?

**Comment by cookandy** [comment]
Hi @aboch.  If I have Tomcat running directly on the host, I don't need to change the MTU on either node - it just works with the default `1500` (even over IPSec).

**Comment by aboch** [comment]
@cookandy Have you tried to specify the 1420 mtu value when creating your network (`--mtu=1420` daemon option for the default `bridge` network, or `--opt com.docker.network.driver.mtu=1420` for your custom network) ?

**Comment by cookandy** [comment]
@aboch Yes, I have.  I've confirmed that the `docker0` interface reflects an MTU of `1420` as well.  If I set the MTU on `docker0`, it doesn't seem to make any difference.

**Comment by cookandy** [comment]
cc: @justincormack 

I did some more digging using Wireshark.  After decrypting the captures, I was able to see the payloads.  

First, with the `eth1` MTU set to `1420` on the requesting machine, and `1500` on the Docker host - everything works:

<img width="1385" alt="screen shot 2016-09-15 at 11 41 11 am" src="https://cloud.githubusercontent.com/assets/6495064/18562619/5c75d142-7b39-11e6-953b-322d903f6d6e.png">

After changing the `eth1` MTU to `1420` on the Docker host, it fails with **[TCP Previous segment not captured]:**

![screen shot 2016-09-15 at 11 44 05 am](https://cloud.githubusercontent.com/assets/6495064/18562717/cc456a1e-7b39-11e6-86d9-c52b09a26dcc.png)

Looking at the failed HTTP payload I can see missing sequence numbers:

![screen shot 2016-09-15 at 11 45 29 am](https://cloud.githubusercontent.com/assets/6495064/18562780/0cf30f58-7b3a-11e6-9263-747228152a2c.png)

Setting the MTU back to `1420` produces the same type of capture.  It's only after I restart `eth1` (resetting the MTU back to `1500`) does it work again:

![screen shot 2016-09-15 at 11 48 43 am](https://cloud.githubusercontent.com/assets/6495064/18562866/6142fec4-7b3a-11e6-90ce-a4a3ec10e011.png)

For added information, here's a capture of me successfully accessing Tomcat **outside of the container** with the MTU set to `1420` on both sides:

![screen shot 2016-09-15 at 11 49 54 am](https://cloud.githubusercontent.com/assets/6495064/18562940/8b7e03be-7b3a-11e6-94b1-c5290468d05b.png)

Is there anything else I can provide to help troubleshoot this further?

**Comment by cookandy** [comment]
Ok, I was finally able to resolve this issue after digging through the bowels of the docker GH issues.

I left all of the MTUs as `1500` and changed it on the firewall:

`iptables -t mangle -A POSTROUTING -p tcp --tcp-flags SYN,RST SYN -o eth1 -j TCPMSS --set-mss 1460 --clamp-mss-to-pmtu`

See here:  http://www.linuxtopia.org/Linux_Firewall_iptables/x4700.html

Other related issues:  

https://github.com/docker/docker/issues/12565#issuecomment-95226447
https://github.com/docker/docker/issues/22028#issuecomment-214004704
https://github.com/docker/docker/issues/22297
https://github.com/docker/docker/issues/3596

**Comment by itwasntandy** [comment]
hey, so you can clamp mss.

However the mistake you were making before I think is reducing the MTU on eth1

you should reduce MTU on the docker0, but leave MTU on eth1 alone. Otherwise all you're doing is reducing the MTU, and not allowing for the IPSEC overhead.
